pkg/observer/observer.go

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright SecureKey Technologies Inc. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package observer
+
+import (
+	"encoding/json"
+
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/core/peer"
+	gossipapi "github.com/hyperledger/fabric/extensions/gossip/api"
+	"github.com/pkg/errors"
+	dcasclient "github.com/trustbloc/fabric-peer-ext/pkg/collections/offledger/dcas/client"
+	"github.com/trustbloc/fabric-peer-ext/pkg/roles"
+	"github.com/trustbloc/sidetree-fabric/pkg/observer/notifier"
+)
+
+var logger = flogging.MustGetLogger("observer")
+
+const (
+	sidetreeNs   = "sidetreetxn_cc"
+	sidetreeColl = "dcas"
+
+	docNs   = "doc_cc"
+	docColl = "diddoc"
+
+	observerRole = "observer"
+)
+
+// getBlockPublisher returns block publisher for channel
+var getBlockPublisher = func(channelID string) publisher {
+	return peer.BlockPublisher.ForChannel(channelID)
+}
+
+// publisher allows clients to add handlers for various block events
+type publisher interface {
+	// AddWriteHandler adds a handler for KV writes
+	AddWriteHandler(handler gossipapi.WriteHandler)
+}
+
+type dcasClient interface {
+
+	// Get retrieves the DCAS value
+	Get(ns, coll, key string) ([]byte, error)
+
+	// Put puts the DCAS value and returns the key for the value
+	Put(ns, coll string, value []byte) (string, error)
+}
+
+type cfg interface {
+	GetChannels() []string
+}
+
+// Start starts observer routines
+func Start(cfg cfg) error {
+
+	if roles.HasRole(observerRole) {
+
+		logger.Infof("peer is an observer, channels to observe: %s", cfg.GetChannels())
+		for _, channelID := range cfg.GetChannels() {
+			initObservingOnChannel(channelID)
+		}
+
+	} else {
+		logger.Info("peer is not an observer, nothing to do...")
+	}
+
+	return nil
+}
+
+func initObservingOnChannel(channelID string) {
```

Comment:
It feels like there should be an Observer struct to encapsulate all of the observer functions, variables.